### ISSUE TITLE
Docker swarm container-container traffic not encrypted when inspecting externally with tcpdump

### ISSUE THREAD

**Issue by stevier5000** [docker/docker#24253]
(NOTE: attempted with both 'testing' and 'experimental' ubuntu repositories
```
osboxes@worker3:~$ docker version
Client:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   906eacd-unsupported
 Built:        Fri Jun 17 21:12:28 2016
 OS/Arch:      linux/amd64
 Experimental: true

Server:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   906eacd-unsupported
 Built:        Fri Jun 17 21:12:28 2016
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**
```
osboxes@worker3:~$ docker info
Containers: 12
 Running: 2
 Paused: 0
 Stopped: 10
Images: 11
Server Version: 1.12.0-rc2
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 66
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay host null bridge
Swarm: active
 NodeID: aat8ei7yd02du6f7qh47aykk5
 IsManager: No
Runtimes: default
Default Runtime: default
Security Options: apparmor seccomp
Kernel Version: 4.4.0-28-generic
Operating System: Ubuntu 16.04 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 2.751 GiB
Name: worker3
ID: 3PJ2:Q4UW:FPJI:URH4:7J2N:ZBTJ:3G54:YOF6:UP35:AIGW:QS7F:HZOJ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Experimental: true
Insecure Registries:
 127.0.0.0/8
osboxes@worker3:~$ docker version
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
VMware fusion VM's. tested with 2 and 3 separate VM's in a swarm

**Steps to reproduce the issue:**
1. docker swarm init; #other host;  docker swarm join ip:2377
2. Create overlay network:
```docker network create -d overlay my-test-network```
3. Deploy docker containers to test. I did a quick and dirty deploy of base ubuntu images, and installed tools within them. 
```docker service create --network my-test-network --replicas 3 --name networktest ubuntu:latest sleep 36000```
4. docker inspect ContainerID to get IP addresses of containers on network
5. Run tcpdump on docker node (i.e. ubuntu VM) or hypervisor (i.e. mac desktop) to watch traffic on port 4789
```sudo tcpdump -s 1500 -A -i ens37 port 4789```
6. Observe that traffic is in cleartext (i.e. install nginx on one container in the service, then install curl on another, and grab the default page. 

**Describe the results you received:**
Traffic when inspected via tcpdump is in cleartext (i.e. http traffic is clearly visible)

**Describe the results you expected:**
Expected to see encapsulated traffic. Not sure if this is expected and I'm not testing this correctly, or what. I.e. do I need two physically separate hosts, mirror a switch port, and watch traffic there, and/or is bridge mode networking within the mac/vmware fusion confusing things?

**Additional information you deem important (e.g. issue happens only occasionally):**

**Comment by mavenugo** [comment]
@stevier5000 Please use the `--opt secure` option when creating a overlay network to enable encryption. Based on the feedback we received, we decided to make it an opt-in.

Please try it out and let us know how it goes.

**Comment by nathanleclaire** [comment]
@mavenugo Is that documented anywhere?  cc @sfsmithcha

**Comment by mavenugo** [comment]
@nathanleclaire am afraid its not.

cc @aboch @thaJeztah

**Comment by stevier5000** [comment]
Confirmed! With that switch on I can't see the traffic in cleartext, it shows up in tcpdump as follows (which makes sense as I seem to remember someone mentioning the secured communications was built on top of ipsec):
```02:51:16.325211 IP worker3 > worker1: esp
02:51:16.325325 IP worker1 > worker3: ESP(spi=0xb80707b5,seq=0x21), length 120
02:51:16.325426 IP worker3 > worker1: ESP(spi=0xb40707b9,seq=0x22), length 1480
02:51:16.325519 IP worker3 > worker1: esp
02:51:16.325572 IP worker3 > worker1: ESP(spi=0xb40707b9,seq=0x23), length 1480
02:51:16.325629 IP worker3 > worker1: esp
02:51:16.325673 IP worker1 > worker3: ESP(spi=0xb80707b5,seq=0x22), length 120
02:51:16.325747 IP worker1 > worker3: ESP(spi=0xb80707b5,seq=0x23), length 120
02:51:16.325796 IP worker3 > worker1: ESP(spi=0xb40707b9,seq=0x24), length 1480```

So, with that bit of information, works great! +1 for documentation stating as much, feel free to close this ticket or turn it into a request for documentation as needed.

**Comment by mavenugo** [comment]
Thanks for confirming @stevier5000 . we will keep this open for doc purposes.

**Comment by bdavidc** [comment]
Has this been added to the doc?

**Comment by thaJeztah** [comment]
@bdavidc I don't think it has been added yet

/cc @sfsmithcha

**Comment by itatabitovski** [comment]
This unfortunately did not work for me. Running in vagrant.

```
docker swarm init --listen-addr 192.168.99.11
docker swarm join --listen-addr 192.168.99.12  --secret 16g0xn6fpiyb8tygej52zv3fg --ca-hash sha256:8518702baac10be6ed9012e9ea5da96ab024ab1fb14263d552402f821c618641 192.168.99.11:2377

docker network create --driver overlay --subnet 10.10.0.0/16 --opt secure mynet

docker service create --name nginx --network mynet --constraint "node.hostname==node-2" -p 8080:80 nginx:1.10-alpine

docker service create --name ping --network mynet --constraint "node.hostname==node-1" alpine ping google.com
```

Trying to connect to nginx service from ping service does not work
```
root@node-1:~# docker exec -ti ping.1.0a8v995iwk406541le59xenq9 sh
/ # ping nginx
PING nginx (10.10.0.2): 56 data bytes
^C
--- nginx ping statistics ---
33 packets transmitted, 0 packets received, 100% packet loss
```

```
root@node-1:~# docker version
Client:
 Version:      1.12.0-rc4
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   e4a0dbc
 Built:        Wed Jul 13 03:35:40 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0-rc4
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   e4a0dbc
 Built:        Wed Jul 13 03:35:40 2016
 OS/Arch:      linux/amd64
```

```
root@node-1:~# docker info
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 1.12.0-rc4
Storage Driver: devicemapper
 Pool Name: docker-8:1-391756-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: ext4
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 313.4 MB
 Data Space Total: 107.4 GB
 Data Space Available: 8.114 GB
 Metadata Space Used: 802.8 kB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.147 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.90 (2014-09-01)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host bridge overlay
Swarm: active
 NodeID: eozz88kihtlvwjnjqv9bl85fd
 IsManager: Yes
 Managers: 1
 Nodes: 2
 CACertHash: sha256:8518702baac10be6ed9012e9ea5da96ab024ab1fb14263d552402f821c618641
Runtimes: runc
Default Runtime: runc
Security Options:
Kernel Version: 4.6.0-0.bpo.1-amd64
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 744.6 MiB
Name: node-1
ID: ANTY:UTOQ:2GOG:QP7F:S7L5:WFCE:B4WQ:UA6Y:OHLS:UXIO:PLK4:ICQW
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No kernel memory limit support
WARNING: No oom kill disable support
Insecure Registries:
 127.0.0.0/8
```

**Comment by thaJeztah** [comment]
@itatabitovski you cannot "ping" services; can you try something else (e.g. `curl`) to connect to the nginx service? See https://github.com/docker/docker/issues/24201 (and https://github.com/docker/docker/issues/23855#issuecomment-228905220)

**Comment by itatabitovski** [comment]
@thaJeztah curl works! thanks.

What I now noticed is that if I have 3 nginx containers/tasks on one node, doing curl nginx does not balance between them.

```
ID                         NAME     SERVICE  IMAGE              LAST STATE                  DESIRED STATE  NODE
7pl9cb6ycpwue4n6zuybdctf9  nginx.1  nginx    nginx:1.11-alpine  Running about a minute ago  Running        node-2
4bo902mgukuaorm8ys2jjxn0c  nginx.2  nginx    nginx:1.11-alpine  Running about a minute ago  Running        node-2
1ric8by8o9w5ggcju0290jfm0  nginx.3  nginx    nginx:1.11-alpine  Running about a minute ago  Running        node-2
```

So ```curl nginx``` only hits one docker container. I expected that the requests are load balanced. Same happens if I curl from the outside on the exposed port.

**Comment by thaJeztah** [comment]
@itatabitovski they are load balanced, but when contacting a service's IP, you're connecting to a "virtual IP" (VIP); not the IP-address of a single container. The Virtual IP doesn't change, but there's a load balancer running on that IP address that makes sure the requests are balanced between the containers.

However, your questions are *not* related to the issue here, which is about encrypting the traffic between containers. It's better not to have this discussion here; if you suspect a bug, open an issue on the issue tracker; for questions about *using* docker, either use forums.docker.com, or the #docker IRC channel.

**Comment by itatabitovski** [comment]
agreed. thank you.

**Comment by thaJeztah** [comment]
docs will be addressed by https://github.com/docker/docker/pull/25135

**Comment by qzio** [comment]
For reference, I just tried this on two aws instances and using `--opt secure` did not seem to encrypt traffic for me since running
`tcpdump -s 1500 -A -i ens37 port 4789` on one of the nodes gave me the plaintext when I tried to communicate between one container on host A and a container on host B using netcat.

However, if I use the `--opt encrypted` flag I did not see any traffic on port 4789 at all but the containers could communicate without trouble.

I did see encrypted traffic using
`tcpdump -i eth0 -n -s 0 -vv \(port 500 or port 4500 or proto 50\)` though.

According to the linked docs, the flag seems to be `--opt encrypted` and not `--opt secure`.
